activerecord/lib/active_record/connection_adapters/postgresql_adapter.rb

```diff
@@ -909,16 +920,24 @@ def add_pg_decoders
             "timestamptz" => PG::TextDecoder::TimestampWithTimeZone,
           }
 
-          known_coder_types = coders_by_name.keys.map { |n| quote(n) }
-          query = <<~SQL % known_coder_types.join(", ")
-            SELECT t.oid, t.typname
-            FROM pg_type as t
-            WHERE t.typname IN (%s)
-          SQL
-          coders = execute_and_clear(query, "SCHEMA", []) do |result|
-            result
-              .map { |row| construct_coder(row, coders_by_name[row["typname"]]) }
-              .compact
+          if schema_cache.known_coder_type_records.present?
```

Comment:
I think I interpreted the feedback a little differently than this. My understanding is that the bulk of this code should end up inside the schema cache, and there will be basically only one line here that fetches the data from the schema cache.

The implementation for this method in the schema cache should be a check to see if the instance variable for this cached value is already present (and if so returning it) similar to the [data_source_exists?](https://github.com/rails/rails/blob/6-1-stable/activerecord/lib/active_record/connection_adapters/schema_cache.rb#L82) method that he linked as reference. If the instance variable doesn't have the cached data, then it should run this query, store the value in the instance variable so its cached for the next time the method is called, and then returning it.

I'm also very interested in getting this merged so let me know if you have additional questions and I would love to help out any way I can

Reply:
I'm excited to see this merged also.

Reply:
@SkipKayhil Thanks for your suggestion, I put `additional_type_records` and `known_coder_type_records` as `class_attribute` because of `additional_type_records` and `known_coder_type_records` should be set before any database connection estiblished. Otherwise, [ActiveRecord::Migrator.current_version](https://github.com/rails/rails/blob/6-1-stable/activerecord/lib/active_record/railtie.rb#L144) would try to estiblish a database connection and invoke the type related queries, which means the cache doesn't work as expected.